Refactor audio handlng

ryze312 opened the pull request:
Hey, here is another PR from me!
Made a few changes to audio handling and cleaned up some of the code. In particular audio devices get started/stopped on demand upon entering voice call. This allows audio backends like PipeWire to suspend sinks when they there isn't anything active connected to them.

`AudioManager`:
* Audio devices get started/stopped on voice connect/disconnect
* Reduce code duplication with `Open/TryOpen` functions
* Don't store device configs and ids

`AudioDevices`:
* Rename `[Get/Set]ActivePlaybackDevice` to `[Get/Set]ActivePlaybackDeviceIter`
* Declare getters as `const` and `[[nodiscard]]`
* Reduce code duplication by using `Get[Playback/Capture]DeviceIDFromModel`
* Add `GetActive[Playback/Capture]`

TODO:
- [x] Start/stop voice audio devices on demand
- [ ] Move notification sounds handling to `AudioManager`
- [ ] Split voice audio into multiple devices per user
- [ ] Add options for persistent audio devices and  split voice audio

ryze312 commented:
The Windows build seems to be failing because of the latest commit involving macOS in master

ouwou commented:
i like. build error will go away if you merge master back in.

for notification sounds: you might need to fiddle with cmake for this (in case notification sounds are enabled and voice is disabled). although maybe u can just change the include guard from WITH_VOICE to WITH_MINIAUDIO. other issue would be how to actually play it considering they use the high-level engine api from miniaudio while voice uses low-level device api. we can just shove both into the AudioManager but if we want to avoid that duplication, we can either use the high-level api for everything incl voice (the engine config can take context and playback device and `ma_engine_read_pcm_frames` can be used in the device callback). or stay with low-level and use something like `ma_decoder_init_file` but idk how well that will work considering eg multiple of same sound at once or repeating it

per-user devices seems niche but im ok with doing that if its not too much of a mess to work with

ouwou commented on src/audio/manager.cpp:
```diff
@@ -130,70 +130,125 @@ AudioManager::AudioManager(const Glib::ustring &backends_string)
-            return;
-        }
+void AudioManager::OpenPlaybackDevice(const ma_device_id device_id) {
+    assert(!m_playback_device_ready && "Tried to open new playback device without closing the current one");
```
i dont use assert anywhere else. would logging a warning not be sufficient ?

ryze312 replied:
I decided to use assert here, since calling Open/Close methods in invalid state would be an error in the code and not an expected behavior. If it's preferable to instead log it, I may do that.

ouwou replied:
my understanding is assert is basically debug-only and ends up calling exit(). i think a message like this is also useful in release builds and even if it is triggered isnt a fatal-force-quit-tier occurrence. so imo i would just log it as an error or warning

ouwou commented on src/audio/manager.cpp:
```diff
@@ -130,70 +130,125 @@ AudioManager::AudioManager(const Glib::ustring &backends_string)
+}
+
+void AudioManager::TryOpenPlaybackDevice(const ma_device_id device_id) {
+    OpenPlaybackDevice(std::move(device_id));
```
std::move doesnt do anything on these structs so they can just be gotten rid of

ryze312 replied:
Not very experienced in C++, I used move here to ensure that the struct doesn't get copied around.

ouwou replied:
it gets copied anyways so it doesnt do anything. also double doesnt do anything because you declare it const (not that it would without the const either). i would just remove them

ryze312 replied:
I suppose I could just pass a reference then to avoid the copy?

ouwou commented on src/audio/manager.cpp:
```diff
@@ -130,70 +130,125 @@ AudioManager::AudioManager(const Glib::ustring &backends_string)
+    OpenPlaybackDevice(std::move(device_id));
+
+    if (m_playback_device_ready) {
+        char name[MA_MAX_DEVICE_NAME_LENGTH + 1];
+        const auto result = ma_device_get_name(&m_playback_device, ma_device_type_playback, name, sizeof(name), nullptr);
+
+        if (result == MA_SUCCESS) {
+            spdlog::get("audio")->info("Started playback device: {}", name);
+        } else {
+            spdlog::get("audio")->info("Started playback device: <unknown>");
+        }
```
```cpp
char name[MA_MAX_DEVICE_NAME_LENGTH + 1] = "<unknown>";
```
then no need for if statement?

ryze312 replied:
I wasn't sure if contents of `name` are valid and safe to use upon error, so I put a branch instead.

ouwou replied:
i checked source and it only writes to buffer on success. im fine with leaving it as is though just an observation

ryze312 commented:
> use the high-level api for everything incl voice

I don't think that's possible, looks like engine API is meant for only for playback. For captures we'll still have to resort to low-level API, especially if we want to support split voice audio, since the audio devices will need to be started/stopped dynamically.

> per-user devices seems niche but im ok with doing that if its not too much of a mess to work with

My main use case for this would be for recording voice calls into separate audio channels in OBS, so that I can manipulate each user's audio independently.

ouwou commented:
> For captures we'll still have to resort to low-level

true. for playback then it might be better to just use engine api and mix device init/callback stuff like i said. capture seems like it would remain the same

> My main use case for this would be for recording voice calls into separate audio channels in OBS, so that I can manipulate each user's audio independently.

ok sure im cool with that. as long as it doesnt clobber up the ui for most people who wont need that

ouwou commented:
uhh i probably figured it was simpler since im not doing any fancy processing and opus's float function just internally converts to s16. rnnoise doesnt actually use f32 it uses s16 also. if you think you get something out of switching over to f32 then go for it

ryze312 commented:
`rnnoise_process_frame`  uses `float*`, the original code converts `int16_t` to `float` and then passes that to the function. Seems strange.
https://github.com/uowuo/abaddon/blob/b3a8356e6bb0c5adf40ca68a4a5daf50ef2ada26/src/audio/manager.cpp#L571-L575

ryze312 commented:
> opus's float function just internally converts to s16

That's only the case if it was compiled with `FIXED_POINT`
```c
#ifdef FIXED_POINT
// snip
#else
opus_int32 opus_encode_float(OpusEncoder *st, const float *pcm, int analysis_frame_size,
                      unsigned char *data, opus_int32 out_data_bytes)
{
   int frame_size;
   frame_size = frame_size_select(analysis_frame_size, st->variable_duration, st->Fs);
   return opus_encode_native(st, pcm, frame_size, data, out_data_bytes, 24,
                             pcm, analysis_frame_size, 0, -2, st->channels, downmix_float, 1);
}
```

https://github.com/xiph/opus/blob/95dbea83486b90256785aa3c75dd2827f591a34c/src/opus_encoder.c#L2522-L2529

ryze312 commented:
Using `s16`, will typically lead to unnecessary conversions:

`Device (f32) -> miniaudio -> Abaddon (s16) -> RNNoise (f32) -> Opus (s16)`

So we might just use `f32` and  `opus_encode_float` and there will be no conversions:


`Device (f32) -> miniaudio -> Abaddon (f32) -> RNNoise (f32) -> Opus (f32)`

ouwou commented:
ok sure lets do f32 then. bonus points for it also being easier to reason with since its easy -1 to 1

ouwou commented:
forgot to mention but i rewrote the buffer stuff just a bit to add a jitter buffer since some ppl were complaining about sound so lemme know if u have questions on how that works. admittedly its not the best implementation

ouwou commented:
just kidding im gonna revert that tomorrow cuz it looks like its causing more problems than it solves

ryze312 commented:
> just kidding im gonna revert that tomorrow cuz it looks like its causing more problems than it solves

Oh okay, I'll see if I can implement it anyway. It looks like the original code uses `std::deque` which is potentially risky in case of overruns/other kinds of blocks, we might want to implement a proper ring buffer

ouwou commented:
you do get underruns but it just results in a little crackle. i encounter it sometimes but its not too bad but i guess its worse if you have bad internet. i think something like libspeexdsp's jitter buffer would be ideal but i dont want to bring it in as a dependency tbh. every jitter buffer i found seems to operate on RTP payloads too so i guess we should also (which my attempt didnt do lol)

ryze312 commented:
> you do get underruns but it just results in a little crackle.

I meant that `FeedMeOpus` just pushes to `std::deque` without any limit to the buffer size. In case of a stall in audio callback, this might lead to delay and/or increase in memory consumption, unless it can catch up by taking frames faster than they get decoded. 

Deque is also a dynamically sized meaning it can cause a reallocation, which is something is generally discouraged in real-time audio processing.

I think we should implement a proper ring buffer with fixed size,  so that Opus decoder stops pushing samples if the audio thread can't keep up.

ouwou commented:
> might lead to delay and/or increase in memory consumption

yeah true. i thought about that but chose to worry about it later 😼. miniaudio has a ring buffer for audio too https://miniaud.io/docs/manual/index.html#RingBuffers
i just wonder whatll happen if it gets congested. the problem with the jitter buffer i wrote was if it filled up too much and started trying to discard data from the start then it would kind of get stuck doing that and the end result is choppy ugly audio. so i wonder if something like that would happen with a ring buffer if it loops back to the front? im sure theres a solution to that but idk it

ryze312 commented:
> miniaudio has a ring buffer for audio too

Haven't seen that, should be easy to integrate

> so i wonder if something like that would happen with a ring buffer if it loops back to the front

I don't think this would happen with ring buffer, since the writer (decoder) is going to discard new frames instead of the old ones in case the reader (audio callback) cannot keep up. The reader can then detect that the buffer is full and and apply something like interpolation  to catch up with the writer. This might distort the audio (higher pitch), but it won't end up choppy.

ouwou commented:
maybe. might be worth looking at how something like mumble does it. i know they use libspeexdsp for jitter so we can see how that interacts

ryze312 commented:
I have split and extracted the voice related functionality of  `AudioManager`  into `VoiceAudio` class that contains `VoicePlayback` and `VoiceCapture`.

I have provided wrappers for some miniaudio structures (`Context`, `AudioDevice`) as well as Opus (`OpusDecoder`, `OpusEncoder`)

Miniaudio structures are wrapped in two layers, classes in `Miniaudio` namespace provide simple wrappers around functions, while `AudioDevice` and `Context` wrap them to fit the use case better (e.g. `AudioDevice` doesn't get invalidated on errors and can change internal device).

`ClientStore` handles creating and removing `Client` objects which meant to represent each RTP peer. They have `VoiceBuffer` as ringbuffer for decoded samples.

`PeakMeter` object handles decaying and raising of respective volume meters (in `VoiceCapture` and `Client`)

Additionally I implemented mutex wrapper (`Mutex`), `std::span`-like slice type (`Slice`) and multi-threaded decoding via channel in thread pool (see: `DecodePool` , `ThreadPool` and `Channel`)

ryze312 commented:
Let me know if I missed anything or you have any questions :3

ryze312 commented:
CI build seems to be failing because of `std::make_optional` in `Context::Create`, not sure why.

Regarding the TODO points. Notifications and other sounds (like someone joining/leaving voice call) can be implemented similar to `VoiceAudio`, make a new class (e. g. `SystemAudio`) that would have `MaEngine` and methods for playing back various sounds. Per use audio device can be an extension to `Client` which would grab audio samples from `VoiceBuffer` in audio callback or write them in `Client::WriteAudio` depending on the setting

ryze312 commented:
That should fix the CI builds

ouwou commented:
ok thats a lot of stuff to look through and im a little burnt out so gimme a bit
CI fail is cuz that particular job has no dependencies enabled so miniaudio isnt available

ryze312 commented:
> im a little burnt out so gimme a bit

Take as much time as you need to, no pressure :3

ouwou commented on src/audio/voice/playback/decode_pool.cpp:
```diff
@@ -0,0 +1,75 @@
+
+namespace AbaddonClient::Audio::Voice::Playback {
+
+// I have no idea what this does, I just copied it from AudioManager
```
rtp payloads can have extension headers which you have to strip off since they arent actually part of the opus packet
https://en.wikipedia.org/wiki/Real-time_Transport_Protocol#Packet_header
now that i think about it the if condition here is probably wrong. oh well thats out of scope here :^)

ouwou commented on src/audio/voice/playback/decode_pool.cpp:
```diff
@@ -0,0 +1,75 @@
+        concurrency = 2;
+    }
+
+    m_pool.m_max_threads = concurrency;
```
should probably getter/setter this

ouwou commented on src/audio/voice/playback/decode_pool.cpp:
```diff
@@ -0,0 +1,75 @@
+}
+
+void DecodePool::DecodeFromRTP(DecodeData &&decode_data) noexcept {
+    m_pool.SendToPool(std::move(decode_data));
```
move also does nothing as long as the vector in the struct is const i think

ouwou commented on src/audio/voice/playback/decode_pool.cpp:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        auto&& decode_message = std::get<DecodeData>(message);
+        OnDecodeMessage(std::move(decode_message));
```
same deal with move

ouwou commented on src/audio/voice/playback/decode_pool.cpp:
```diff
@@ -0,0 +1,75 @@
+    }
+}
+
+void DecodePool::OnDecodeMessage(DecodeData &&message) noexcept {
+    auto&& [rtp, decoder, buffer] = message;
```
can just make this a standard reference?

ouwou commented on src/audio/voice/playback/voice_buffer.hpp:
```diff
@@ -0,0 +1,27 @@
+    uint32_t m_channels;
+    uint32_t m_buffer_frames;
+
+    bool moved = false;
```
`m_` for consistencys sake

ouwou commented on src/audio/voice/playback/voice_playback.cpp:
```diff
@@ -0,0 +1,83 @@
+namespace AbaddonClient::Audio::Voice {
+
+void playback_callback(ma_device* pDevice, void* pOutput, const void* pInput, ma_uint32 frameCount) {
+    auto playback = reinterpret_cast<VoicePlayback*>(pDevice->pUserData);
```
also can be static_cast

ouwou commented on src/audio/voice/playback/voice_playback.cpp:
```diff
@@ -0,0 +1,83 @@
+
+void VoicePlayback::OnRTPData(ClientID id, const std::vector<uint8_t> &&data) noexcept {
+    if (m_active) {
+        m_clients.DecodeFromRTP(id, std::move(data));
```
move does nothing

ouwou commented on src/audio/voice/playback/voice_playback.cpp:
```diff
@@ -0,0 +1,83 @@
+void VoicePlayback::SetPlaybackDevice(ma_device_id &&device_id) noexcept {
+    spdlog::get("voice")->info("Setting playback device");
+
+    const auto success = m_device.ChangeDevice(std::move(device_id));
```
move does nothing

ouwou commented on src/components/channellist/channellisttree.cpp:
```diff
@@ -636,6 +636,12 @@ void ChannelListTree::OnVoiceUserDisconnect(Snowflake user_id, Snowflake channel
     if (auto iter = GetIteratorForRowFromIDOfType(user_id, RenderType::VoiceParticipant)) {
         m_model->erase(iter);
     }
+
+    auto& abaddon = Abaddon::Get();
+    auto ssrc = abaddon.GetDiscordClient().GetSSRCOfUser(user_id);
+    if (ssrc) {
+        abaddon.GetAudio().RemoveSSRC(*ssrc);
+    }
```
not the right place to do this. the signal is emitted here:
https://github.com/uowuo/abaddon/blob/ada3420dce68a3f4f5a3f1bcaf55f8d688b3d27f/src/discord/discord.cpp#L2389-L2404
if we wanna avoid duplication maybe we can split out another function in DiscordClient for disconnects but i dont think (effectively) 2 lines of duplication is a big deal

ryze312 replied:
I suppose it would be better to instead make `VoiceAudio` connect to these signals and handle them there? For now I just redirect methods from `AudioManager` to stay compatible.

ouwou replied:
connecting there is probably fine. cant think of a better place off top of my head just not in the ui code lul

ouwou commented on src/misc/mutex.hpp:
```diff
@@ -0,0 +1,101 @@
+class Mutex {
+
+public:
+    Mutex() {}
```
```suggestion
    Mutex() = default;
```

ryze312 replied:
Not sure why I left that in, might as well remove it entirely, since it's implicit anyway.

ouwou commented on src/misc/threadpool.hpp:
```diff
@@ -0,0 +1,68 @@
+public:
+    using ThreadMessage = std::variant<ThreadData, Terminate>;
+
+    ThreadPool() noexcept;
```
```suggestion
    ThreadPool() noexcept = default;
```

ouwou commented on src/misc/threadpool.hpp:
```diff
@@ -0,0 +1,68 @@
+        m_channel.Send(std::move(data));
+    }
+
+    size_t m_max_threads;
```
this can be uninitialized. maybe just move to constructor? also private

ouwou commented on src/audio/voice/capture/voice_capture.cpp:
```diff
@@ -0,0 +1,165 @@
+    opus.resize(bytes);
+
+    if (bytes > 0) {
+        m_signal.emit(std::move(opus));
```
emit will effectively take a const reference here so move does nothing

ouwou replied:
i dont think sigc will do any copies if the connected slot takes it as a reference

ryze312 replied:
> i dont think sigc will do any copies

This wouldn't make sense, as the vector would get  destroyed at the end of the scope there. I tried taking `std:vector<uint8_t>&&` in the slot, but couldn't make it compile.

ouwou replied:
u can just take a normal const reference. all slots are called when you do emit so itll be in scope

ryze312 replied:
My understanding is that sigc accepts the parameters and passes them down to all slots without waiting for callbacks to return (thus non-blocking). If that is the case, stack will get corrupted if `EncodeAndSend` scope ends before the callback returns (as it will still hold the reference to the vector). 

We need to either copy or ideally move the vector here for safe behavior.

ouwou replied:
not sure where you got that from but sigc definitely waits. what youre describing is more like how Glib::Dispatcher works. emit also explicitly returns whatever the last connected slot returned

ryze312 replied:
Oh so it is blocking?  I couldn't find much information about it. Using `const&` should be fine then.

ouwou commented on src/audio/voice/capture/voice_capture.cpp:
```diff
@@ -0,0 +1,165 @@
+using EncodingApplication = OpusEncoder::EncodingApplication;
+
+void capture_callback(ma_device* pDevice, void* pOutput, const void* pInput, ma_uint32 frameCount) {
+    auto capture = reinterpret_cast<VoiceCapture*>(pDevice->pUserData);
```
you probably got this from me cuz i didnt know but this can be a static_cast

ryze312 replied:
> you probably got this from me

Yes I did

ouwou reviewed (commented):
i just took a scroll through so this is really surface level. theres a lot of abstraction here so i still need to wrap my head around all that.
also code style is obviously a lot different here (probably in a good way lol). the stuff like ConstSlice is a little weird to me but ill probably get used to it
good stuff tho. now i still need to test :^)

ouwou commented on src/audio/utils.hpp:
```diff
@@ -0,0 +1,40 @@
+
+    static void MixStereoToMono(OutputBuffer buffer) noexcept {
+        for (auto iter = buffer.begin(); iter < buffer.end() - 2; iter += 2) {
+            const auto mixed = std::reduce(iter, iter + 2, 0) / 2.0f;
```
```suggestion
            const auto mixed = std::reduce(iter, iter + 2, 0.0f) / 2.0f;
```
reduce is deducing as int here so mix doesnt work. unrelated but it looks like discord is mixing down to mono on the receiving end too now. >:^(

ouwou commented:
another thing that im pretty sure worked was if you set someones volume and they rejoin or you rejoin the volume is preserved which doesnt seem to work anymore. ideally mute would also carry over as well

ryze312 commented on src/audio/manager.cpp:
```diff
@@ -684,36 +483,36 @@ std::vector<ma_backend> AudioManager::ParseBackendsList(const Glib::ustring &lis
 
 double AudioManager::GetRNNProbThreshold() const {
-    return m_prob_threshold;
+    return m_voice->GetCapture().GetEffects().GetNoise().m_vad_threshold;
```
I see you reference the same thing later, I am gonna say this only here, but the same applies to all other occurrences.

I am not sure getters and setters make sense here, since they would effectively do the same thing, but the calls won't be able to get optimized away unless you build with LTO (which iirc is not the case for your release builds).

ouwou replied:
i dont think optimizing away a function call is of any practical importance but if we wanna stick with it then we should call it something like `VADThreshold`. i just prefer get/set as a matter of style

ryze312 replied:
Yeah I am happy with that. Get/Set is ambiguous to the caller in what it would do, might just be a simple field or something more expensive.

ryze312 commented on src/audio/miniaudio/ma_context.cpp:
```diff
@@ -0,0 +1,43 @@
+        return std::nullopt;
+    }
+
+    return std::make_optional<MaContext>(std::move(context));
```
You want it to be like this?
```cpp
return MaContext(std::move(context));
```

`std::make_optional` constructs the value in-place so that saves a move

ouwou replied:
you mean as a result of it being returned? im fairly certain this would behave exactly the same:
```cpp
return std::move(context);
```

ryze312 replied:
Looks like it does, will go with that

ryze312 commented on src/audio/voice/capture/voice_capture.cpp:
```diff
@@ -0,0 +1,165 @@
+    m_active = active;
+}
+
+void VoiceCapture::SetCaptureDevice(ma_device_id &&device_id) noexcept {
+    spdlog::get("voice")->info("Setting capture device");
+
+    const auto success = m_device.ChangeDevice(std::move(device_id));
```
That's true, should we use const reference instead to avoid copying?

ouwou replied:
might as well. i use const& as the default basically

ryze312 commented on src/audio/voice/capture/voice_capture.cpp:
```diff
@@ -0,0 +1,165 @@
+
+    ApplyEffects(buffer);
+    if (ApplyNoise(buffer)) {
+        EncodeAndSend(std::move(buffer));
```
I suppose taking `CaptureBuffer&` would be better to avoid copy, I was trying to use rvalue reference to express ownership, but looks like that just copies `std::array`

ouwou replied:
yeah `CaptureBuffer&` is totally fine. array is basically just a wrapper over a c-style array so theres nothing for a move to steal from

ryze312 commented:
> the volume is preserved which doesnt seem to work anymore

This seems to be a side of effect of `RemoveSSRC`, since it effectively removes the client along with volume value. I think the volume should be stored separately in this case, so it persists even if we rejoin the call too.

ryze312 commented:
> the stuff like ConstSlice is a little weird to me

It is effectively the same container as [`std::span`](https://en.cppreference.com/w/cpp/container/span) since C++20 (thus I couldn't use it, the cause codebase is C++17). I called it `Slice`, since it holds size and pointer as just like [`slice` in Rust](https://doc.rust-lang.org/std/primitive.slice.html)

ouwou commented:
> I think the volume should be stored separately in this case, so it persists even if we rejoin the call too.

it does albeit in maybe a weird place now
https://github.com/uowuo/abaddon/blob/cba2f6c66ef27c992072afa79ebaae541814506b/src/discord/voiceclient.cpp#L230-L235
it was only used for the ui though. u can see the volume slider value is preserved across disconnected. theres probably a smarter way to do it now though

ryze312 commented:
> it does albeit in maybe a weird place now

I see, I suppose we can could that for that to restore the volume value in `Client` for now

ryze312 commented:
If you are happy with everything so far, I can go ahead and simplify `AudioManager` and just make it expose `VoiceAudio`. After that I can do `SystemAudio` for notification/voice chat sounds by the same approach.

ouwou commented:
sure sounds good !

ryze312 commented:
I have no idea why, but it looks like `signal_voice_user_connect` and `signal_voice_user_disconnect` get emitted at seemingly random times during the call.

ouwou commented on src/audio/system/system_audio.cpp:
```diff
@@ -0,0 +1,66 @@
+}
+
+void SystemAudio::OnVoiceUserConnect(Snowflake user_id, Snowflake channel_id) noexcept {
+    PlaySound(SystemSound::VoiceConnected);
```
im guessing your comment is because the sound is playing a lot? this signal is emitted for all channels so you prob wanna check for channel_id to match the current voice channel (`DiscordClient::GetVoiceChannelID`)

ouwou commented on src/discord/voiceclient.cpp:
```diff
@@ -152,9 +151,8 @@ DiscordVoiceClient::DiscordVoiceClient()
-        audio.SetOpusBuffer(m_opus_buffer.data());
-        audio.signal_opus_packet().connect([this](const std::vector<uint8_t> opus) {
+        auto &capture = Abaddon::Get().GetAudio().GetVoice().GetCapture();
+        capture.GetCaptureSignal().connect([this](const std::vector<uint8_t> opus) {
```
`const std::vector<uint8_t> &opus` should avoid copy

ouwou commented on src/audio/audio_engine.cpp:
```diff
@@ -0,0 +1,50 @@
+void AudioEngine::StartTimer() noexcept {
+    // NOTE: I am not using g_timeout_add_seconds_once here since we want to own the source and destroy it manually
+    // g_source_remove throws an error on destroyed source
+    m_timer_source = g_timeout_source_new_seconds(5);
```
im guessing this is meant to kill the engine if no sounds are played for 5 seconds? might it just be simpler to keep it around anyways? or maybe we can just call `ma_engine_stop` to avoid having to create it over and over.
also im pretty sure there are glibmm bindings for this but thats not a big deal

ryze312 replied:
I was trying to match the behavior of the official client, it spawns source as needed with the timeout of 5 seconds.  `ma_engine_stop` seems to just stop the device's callbacks internally, it still keeps the source intact and connected.

glibmm has `SignalTimeout::connect_seconds_once`, but it doesn't seem to provide a way to destroy/reset the timeout.

ouwou commented:
maybe we can come up with better names for SystemAudio and AudioEngine since theyre a little confusing since system audio is about event sounds and audio engine is meant for playing sound files. idk if AudioEngine even really needs to be its own thing

ryze312 commented:
I think this should be all

ouwou commented:
still need to fix the build when compiling without audio support so a bunch of stuff probably needs to be wrapped in `#ifdef WITH_MINIAUDIO`
ill give everything another test soon ™

ryze312 commented:
> a bunch of stuff probably needs to be wrapped in `#ifdef WITH_MINIAUDIO`

I think the right approach would be not including the sources in CMake at all, not only does it increase the compilation time, but also having to include the guard in each file seems tedious.

Just did that, seems to compile and work fine.

ryze312 commented:
Uh nevermind

ryze312 commented:
The build seems to be failing now because of  LTO. Might be [this issue](https://github.com/msys2/MINGW-packages/issues/11726)?

ouwou commented:
maybe this? https://github.com/diasurgical/devilutionX/pull/6656/commits/8693a0f66ff7d1003ce19e7271d3f832ec7135aa

ryze312 commented:
Sure, I'll see if disabling decloning for those specific files works, if not, we can try disabling it for the entire executable.

ryze312 commented:
I have no idea what's breaking it now, I'll just disable LTO for Windows

ouwou commented:
theres a good bit of missing headers. it builds without them thanks to precompiled headers meaning everything ends up transiently included but they should be added anyways (it breaks my intellisense too). you can disable precompiled headers in cmake and it should give you errors where stuff is missing. adding headers wont hurt compile time (at least in my testing) if they are included in the precompiled headers

ouwou commented:
ok giving it some more testing

1. suppress noise seems to only denoise the right channel
2. `StripRTPExtensionHeader` is broken. obviously this isnt ur fault but if i fix it in master itll cause a conflict here so it should probably be fixed here. i think i can add a pretty easy fix so ill comment a diff soon-ish
3. i get a segfault eventually at mutex.hpp:93 🙀
```
#0  0x00007ffbf8172f58 in ?? () from C:\msys64\mingw64\bin\libwinpthread-1.dll
#1  0x00007ff79f6b1f1e in __gthread_mutex_lock (__mutex=0xbaadf00dbaadf00d) at C:/msys64/mingw64/include/c++/14.1.0/x86_64-w64-mingw32/bits/gthr-default.h:762
#2  std::mutex::lock (this=0xbaadf00dbaadf00d) at C:/msys64/mingw64/include/c++/14.1.0/bits/std_mutex.h:113
#3  std::unique_lock<std::mutex>::lock (this=<synthetic pointer>) at C:/msys64/mingw64/include/c++/14.1.0/bits/unique_lock.h:147
#4  std::unique_lock<std::mutex>::unique_lock (this=<synthetic pointer>, __m=...) at C:/msys64/mingw64/include/c++/14.1.0/bits/unique_lock.h:73
#5  Mutex<AbaddonClient::Audio::Voice::Opus::OpusDecoder>::Lock (this=0xbaadf00dbaadf00d) at C:/path/src/misc/mutex.hpp:93
#6  AbaddonClient::Audio::Voice::Playback::DecodePool::OnDecodeMessage (message=...) at C:/path/src/audio/voice/playback/decode_pool.cpp:54
#7  0x00007ff79f6b20b1 in AbaddonClient::Audio::Voice::Playback::DecodePool::DecodeThread (channel=...) at C:/path/src/audio/voice/playback/decode_pool.cpp:44
#8  0x00007ffb72fc578f in ?? () from C:\msys64\mingw64\bin\libstdc++-6.dll
#9  0x00007ffbf8174dbb in ?? () from C:\msys64\mingw64\bin\libwinpthread-1.dll
#10 0x00007ffc098eaf5a in msvcrt!_beginthreadex () from C:\Windows\System32\msvcrt.dll
#11 0x00007ffc098eb02c in msvcrt!_endthreadex () from C:\Windows\System32\msvcrt.dll
#12 0x00007ffc09f07344 in KERNEL32!BaseThreadInitThunk () from C:\Windows\System32\kernel32.dll
#13 0x00007ffc0b8bcc91 in ntdll!RtlUserThreadStart () from C:\Windows\SYSTEM32\ntdll.dll
#14 0x0000000000000000 in ?? ()
```

ouwou commented:
e6191d95341d164eacaf91739c8aa7020dd5c9b6 should address point 2. shouldnt be too hard to port over here

ryze312 commented:
> suppress noise seems to only denoise the right channel
Seems to happen if you use combination of RNNoise as VAD and then switch it to gate.

ryze312 commented:
>  i get a segfault eventually at mutex.hpp:93

Not sure what could be wrong here, `shared_ptr` should prevent `Mutex` from being freed
